### DISCUSSION TITLE
Inject dependencies in interceptors constructors

### DISCUSSION THREAD

**Discussion by facuabdo** [lawrence-laz/Decor.NET#9] [Q&A]
Hi Lawrence, i found your package of great utility. I was wondering if dependencies that need to be injected into the intercerptors using the constructor are considered -in a way that they're resolved automatically- or if you need to resolve them manually.

**Comment by lawrence-laz** [comment]
Hi facuabdo, I'm glad you find the package helpful. 🙂

Yes, automatic dependency injection is considered for interceptor classes. There are two ways of achieving this.

1.  If you are using Microsoft's `IServiceProvider` as a dependency container, then you can use this package [Decor.Extensions.Microsoft.DependencyInjection](https://www.nuget.org/packages/Decor.Extensions.Microsoft.DependencyInjection/) and follow these [instructions](https://github.com/lawrence-laz/Decor.NET#with-microsoft-dependency-injection).

2. If you use some other dependency container, then you can provide a [Factory](https://github.com/lawrence-laz/Decor.NET/blob/341b87140f69a20a73f0790ff8b42e4a585d1c5f/Decor/Decorator.cs#L19) method with your container's resolve implementation. Pseudo example:
    ```csharp
    Decorator decorator = ... ;
    YourDependencyContainer container = ... ;
    decorator.Factory = (type) => (IDecorator)container.Resolve(type);
    ```
    After this the `Decorator` class will be able to resolve constructor dependencies automatically.

Let me know if this helps. Otherwise I could provide a minimal working example if you share which dependency container you are using.

**Reply by facuabdo** [reply]
Thanks for the reply, it definitely helps. I missed mentioning i was using ASP.NET Core built-in DI. While I was waiting for your reply I ran some tests and found out it works out of the box (all dependencies are resolved). 

One thing that is getting me some trouble is that my interceptor has a DbContext dependency. The only way I can get it to work is registering DbContext as transient, which is something I want to avoid since i'm trying to implement Unit of Work using interceptors + DbContext.

This is my ServiceCommandInterceptor code 

```
public class ServiceCommandInterceptor : IDecorator
    {
        private readonly DbContext _unitOfWork;

        public ServiceCommandInterceptor(DbContext unitOfWork)
        {
            _unitOfWork = unitOfWork;
        }

        public Task OnInvoke(Call call)
        {
            using (IDbContextTransaction transaction = _unitOfWork.Database.BeginTransaction())
            {
                try
                {
                    var result = call.Next();
                    _unitOfWork.SaveChanges();
                    transaction.Commit();
                    return result;
                }
                catch
                {
                    transaction.Rollback();
                    throw;
                }
            }
        }
    }
```
And this are my registrations in the ConfigureServices method

```
public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers();

            
            services.AddScoped<DbContext, InmobiliariaCloudDbContext>();
            services.AddScoped(typeof(IRepository<>), typeof(BaseEFRepository<>));

            services.AddScoped<IAccountService, AccountService>();
            services.AddTransient<ServiceCommandInterceptor>();
            services.Decorate<IAccountService>();

            services.AddScoped<IAccountService, AccountService>();
            services.AddTransient<ServiceQueryInterceptor>();
            services.Decorate<IAccountService>();
        }
```

Calling any method from AccountService throws System.InvalidOperationException: Cannot resolve 'Services.Core.ServiceCommandInterceptor' from root provider because it requires scoped service 'Microsoft.EntityFrameworkCore.DbContext'.

Changing 1st line to

```services.AddTransient<DbContext, InmobiliariaCloudDbContext>();```

Works just fine. 

In this case my AccountService has only one depedency on IRepository<Account> which is resolved to BaseEFRepository<Account>. BaseEFRepository<T> as well has a dependency on DbContext. If my service had more dependencies on more repositories it would mean that each of them would receive a new instance of DbContext (as a result of being registered as transient) and unit of work would not work at all. Any ideas why it can't be resolved when registered as scoped?

PS: sorry for the many many edits, i couldn't get the code to display correctly

**Reply by lawrence-laz** [reply]
I was pretty confident scoped dependencies in transient `IDecorator` shouldn't cause any problems, but [added this test](https://github.com/lawrence-laz/Decor.NET/commit/a0c0903190b36f69ebe61426e2ee34094c2392b7) just to make sure. The test passes without problems.

The error seems like it could be specific to Asp.NET Core environment. Could you possibly provide me with a minimal part of your project that reproduces this problem?

Oh and I noticed a slight unrelated problem in your `OnInvoke` method: `call.Next()` call should be awaited. Otherwise `_unitOfWork.SaveChanges();` might get executed before the service finishes its work.
